templates/vagrant/.vagrantplugins.erb

```diff
@@ -1,7 +1,23 @@
-<% if plugins.any? -%>
-required_plugins = %w( <% plugins.each do |p| -%><%= p %> <% end -%>)
+required_plugins = {}
+# Example usage:
+# required_plugins["plugin-name"] = { version: "1.2.3", source: "https://rubygems.org" }
+<% plugins.each do |p| -%>
+required_plugins["<%= p %>"] = {}
+<% end -%>
 
-required_plugins.each do |plugin|
-  system "vagrant plugin install #{plugin}" unless Vagrant.has_plugin? plugin
+needs_restart = false
+required_plugins.each do |plugin, options|
+  version = options[:version]
+
+  unless Vagrant.has_plugin?(plugin, version)
+    command = "vagrant plugin install #{plugin}"
+    command += " --plugin-version #{version}" if version
+    command += " --plugin-source #{options[:source]}" if options[:source]
+    system command
+    needs_restart = true
+  end
+end
+
+if needs_restart
```

Comment:
In my very simple testing, it seems to do the right thing. I'll spend a little more time on the edges.